File: Heap/Python/Maximum Subsequence Score.py
# Problem link: https://leetcode.com/problems/maximum-subsequence-score/description/
# Time: O(nlogn)
# Space: O(n)

class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        store, minHeap, res, total = [], [], -inf, 0

        for i in range(len(nums1)):
            store.append((nums2[i], nums1[i]))
        
        store.sort(reverse = True)

        for p1, p2 in store:
            heapq.heappush(minHeap, p2)
            total += p2
            
            if len(minHeap) == k:
                res = max(res, p1 * total)
                total -= heapq.heappop(minHeap)
        
        return res

File: Heap/Java/Maximum Subsequence Score.java
// Problem link: https://leetcode.com/problems/maximum-subsequence-score/description/
// Time: O(nlogn)
// Space: O(n)
class Solution {
    public long maxScore(int[] nums1, int[] nums2, int k) {
        List<int[]> store = new ArrayList<>();
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        long res = Integer.MIN_VALUE;
        long total = 0;

        for (int i = 0; i < nums1.length; i++) {
            store.add(new int[]{nums2[i], nums1[i]});
        }

        store.sort((a, b) -> Integer.compare(b[0], a[0]));

        for (int[] pair : store) {
            minHeap.offer(pair[1]);
            total += pair[1];

            if (minHeap.size() == k) {
                res = Math.max(res, pair[0] * total);
                total -= minHeap.poll();
            }
        }

        return res;
    }
}
